# Cleanup on Row 7

1. Clean the File
2. Handle Corrupt Rows
3. Explore Data Set Bias

## Import Libraries

In [54]:
import numpy as np
import pandas as pd
from datetime import datetime
import matplotlib
import matplotlib.pyplot as plt

### Helper Display Functions

In [3]:
def display_df(df, nrows=10, ncols=None):
    with pd.option_context('display.max_rows', nrows, 'display.max_columns', ncols):
        display (df)
def print_row(df, row):
    for ctr,i in enumerate(df.iloc[row]):
        print (str(df.columns[ctr])+": "+str(i))

### Helper Statistical Functions

In [4]:
#gets ratio of NaNs for each column
def stats_NaN(df):
    df_stats = pd.DataFrame(index=[df.columns], columns=["NaN Ratio"])
    for col in df.columns:
        df_stats["NaN Ratio"][col] = df[col].isna().sum()/len(df) #NaN ratio
    return df_stats.sort_values(by=['NaN Ratio'])

#helps see what fields may be incorrect by having more values than should be
def stats_unique(df, labels):
    df_counts = pd.DataFrame(index=labels,columns=['Unique Count'])
    for l in labels:
        df_counts['Unique Count'][l] = df[l].nunique()
    return df_counts

## Cleaning the Data

### Importing the Data

In [5]:
df_raw = pd.read_csv("../dirty_sample_small.csv")

ParserError: Error tokenizing data. C error: Expected 48 fields in line 652242, saw 49


Some rows have an extra field. We shall investigate what this is. First we will import the file in a more managable way, capable of handling these erroneous lines.

In [10]:
df_raw = pd.DataFrame([line.strip().split(',') for line in open('../dirty_sample_small.csv', 
                                                                'r', encoding="ISO-8859-1")])

In [11]:
len(df_raw)

661487

### Investigating Corrupt Lines (Part 1)

One notion of corrupt lines is the lines in the raw dataset that are formatted improperly. These lines have made it difficult to import the file, though we have worked around it with the above code. These corrupt lines have to do with cleaning the dataset so will discuss it here. Interestingly, we will check some statistics on these lines with regards to duplicate entires. There is another notion of corrupt lines, dealing with the incorrectness and contradictoriness of datapoints, that we will investigate later.

In [12]:
stats_NaN(df_raw)

NaN Ratio
0             0
1             0
2             0
3   0.000270602
4   0.000498876
5   0.000757384
6    0.00102496
7    0.00126231
8    0.00155407
9    0.00181712
10   0.00207563
11   0.00235379
12   0.00264404
13   0.00291767
14   0.00316257
15   0.00343166
16   0.00372192
17   0.00397589
18   0.00421928
19   0.00446116
20   0.00468339
21   0.00495399
22   0.00520948
23   0.00549066
24   0.00576882
25   0.00600616
26   0.00624804
27   0.00652016
28   0.00674238
29   0.00698426
30   0.00722463
31   0.00749977
32   0.00774618
33   0.00801225
34   0.00826471
35   0.00856706
36   0.00882406
37   0.00909617
38   0.00936526
39   0.00962075
40   0.00989891
41   0.00989891
42   0.00989891
43   0.00989891
44   0.00989891
45   0.00989891
46   0.00989891
47   0.00989891
48     0.986152

We see that that the extra has 98.6% field NaNs. This is enough to drop it. However we want to do our due diligence and investigate what the values that are not NaN are. Our suspicion is that it contains the empty string "". So we will query this dataset on the extra field for values that are not NaN or not the empty string. These rows with the empty string are likely corrupt.

In [16]:
df_raw[(pd.notna(df_raw[48]) & (df_raw[48] != ''))][48]

Series([], Name: 48, dtype: object)

We see that this returns nothing. Therefore everything in the extra field is either NaN or the empty string, so there is absolutely no value in keeping it and we will drop it. But first, lets see how many of these corrupt erroneous lines there are. We know that the construction of df_raw line by line as we did above is indifferent to the amount of columns in each row. Most of the time it saw 48 columns, in the later rows it saw 49 because of the extra empty string. The construction of the data set put NaNs in the earlier rows to ensure the shape of the dataset is intact (a regular rectangle). Therefore the source of the corrupt lines is the empty string, so if we count the empty strings in the extra field we will have successfully counted the amount of corrupt lines.

In [17]:
len(df_raw[((df_raw[48] == ''))][48])

9160

We see that there are 9160 of these rows.

In [18]:
display_df(df_raw[((df_raw[48] == ''))])

0        1     2      3     4      5      6   \
652241  HarvardX/CS50x/2014_T1  1818882  True   True  True  False  False   
652242  HarvardX/CS50x/2014_T1  1818882  True   True  True  False  False   
652243  HarvardX/CS50x/2014_T1  1818882  True   True  True  False  False   
652244  HarvardX/CS50x/2014_T1  1818882  True   True  True  False  False   
652245  HarvardX/CS50x/2014_T1  1818882  True   True  True  False  False   
...                        ...      ...   ...    ...   ...    ...    ...   
661482  HarvardX/CS50x/2014_T1  1934738  True  False        False  False   
661483  HarvardX/CS50x/2014_T1  1934738  True  False        False  False   
661484  HarvardX/CS50x/2014_T1  1934738  True  False        False  False   
661485  HarvardX/CS50x/2014_T1  1934738  True  False        False  False   
661486  HarvardX/CS50x/2014_T1  1934738  True  False        False  False   

                    7   8              9         10      11  12  \
652241  173.76.227.129  US  United States  Americas  Newton  MA   
652242  173.76.227.129  US  United States  Americas  Newton  MA   
652243  173.76.227.129  US  United States  Americas  Newton  MA   
652244  173.76.227.129  US  United States  Americas  Newton  MA   
652245  173.76.227.129  US  United States  Americas  Newton  MA   
...                ...  ..            ...       ...     ...  ..   
661482   71.41.229.179  US  United States  Americas  Dallas  TX   
661483   71.41.229.179  US  United States  Americas  Dallas  TX   
661484   71.41.229.179  US  United States  Americas  Dallas  TX   
661485   71.41.229.179  US  United States  Americas  Dallas  TX   
661486   71.41.229.179  US  United States  Americas  Dallas  TX   

                   13     14                15                 16 17 18  \
652241  Massachusetts  02458  Northern America  Developed regions         
652242  Massachusetts  02458  Northern America  Developed regions         
652243  Massachusetts  02458  Northern America  Developed regions         
652244  Massachusetts  02458  Northern America  Developed regions         
652245  Massachusetts  02458  Northern America  Developed regions         
...               ...    ...               ...                ... .. ..   
661482          Texas  75201  Northern America  Developed regions         
661483          Texas  75201  Northern America  Developed regions         
661484          Texas  75201  Northern America  Developed regions         
661485          Texas  75201  Northern America  Developed regions         
661486          Texas  75201  Northern America  Developed regions         

             19        20  21    22 23   24   25                   26  \
652241  42.3533  -71.1883  el  2002  m  0.0  0.6  2013-12-04 04:02:20   
652242  42.3533  -71.1883  el  2002  m  0.0  0.6  2013-12-04 04:02:20   
652243  42.3533  -71.1883  el  2002  m  0.0  0.6  2013-12-04 04:02:20   
652244  42.3533  -71.1883  el  2002  m  0.0  0.6  2013-12-04 04:02:20   
652245  42.3533  -71.1883  el  2002  m  0.0  0.6  2013-12-04 04:02:20   
...         ...       ...  ..   ... ..  ...  ...                  ...   
661482  32.7831  -96.8067   p  1972  m  0.0  0.6  2013-07-09 01:46:41   
661483  32.7831  -96.8067   p  1972  m  0.0  0.6  2013-07-09 01:46:41   
661484  32.7831  -96.8067   p  1972  m  0.0  0.6  2013-07-09 01:46:41   
661485  32.7831  -96.8067   p  1972  m  0.0  0.6  2013-07-09 01:46:41   
661486  32.7831  -96.8067   p  1972  m  0.0  0.6  2013-07-09 01:46:41   

                                27                          28    29  30   31  \
652241  2013-12-04 04:02:20.199004  2014-04-01 23:30:47.988143  2194  39  680   
652242  2013-12-04 04:02:20.199004  2014-04-01 23:30:47.988143  2194  39  680   
652243  2013-12-04 04:02:20.199004  2014-04-01 23:30:47.988143  2194  39  680   
652244  2013-12-04 04:02:20.199004  2014-04-01 23:30:47.988143  2194  39  680   
652245  2013-12-04 04:02:20.199004  2014-04-01 23:30:47.988143  2194  39  680   
...                            ...            

Many of these "corrupt" lines contains valid data, so we will have to handle them. It does appear though that many of these 9160 corrupt lines are indeed duplicates. Lets see how many.

In [19]:
len(df_raw[((df_raw[48] == ''))].drop_duplicates(inplace=False))

710

So of the 9160 corrupt lines, at least 9160 - 710 = 8450 are duplicates. I say at least because even those 710 lines may be duplicates with the rest of the raw dataset as this analysis above only considers rows with the empty string in the extra field. So we can analyze overall the amount of duplicates in the whole raw dataset and compare.

In [20]:
len(df_raw.drop_duplicates(inplace=False))

57223

Interestingly, there are still many more duplicates. However, if we had dropped duplicates at this stage, then we would have also gotten rid of at least 8450 corrupt lines. Therefore we posit, that dropping duplicates would drastically change the count of corrupt lines. We shall prove this with the code below.

In [21]:
len(df_raw.drop_duplicates(inplace=False)[df_raw[48] == ''])

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


710

We see that that these corrupt lines do have 710 valid samples. Therefore simply dropping duplicates in the whole dataset is guarunteed to drop 8450 of the corrupt lines.

### Fixing the Shifted Data

We observe that some of the data appears scrambled. Meaning some of the data for some columns is in the wrong columns. We elucidate below. But first, lets give our dataset acceptable column names and drop the erroneous column. From now on, we will be using the goal dataset (the future fully cleaned one) called `df_clean`.

In [453]:
df_clean = df_raw.copy()
df_clean = df_clean.rename(columns=df_raw.iloc[0])
df_clean = df_clean.drop(df_clean.index[0])
df_clean = df_clean.drop(np.nan, axis=1)

In [337]:
stats_unique(df_clean, df_clean.columns)

Unique Count
course_id                      21
user_id                     49144
registered                      3
viewed                          3
explored                        3
certified                       2
completed                   44101
ip                            799
cc_by_ip                      295
countryLabel                   96
continent                    6509
city                         1145
region                       1646
subdivision                  9055
postalCode                    282
un_major_region                21
un_economic_group               4
un_developing_nation            4
un_special_region            9821
latitude                    10569
longitude                     443
LoE                           115
YoB                            59
gender                        105
grade                           8
passing_grade               47896
start_time                  33443
first_event                 40371
last_event                   3238
nevents                       267
ndays_act                     873
nplay_video                   111
nchapters                      68
nforum_posts                   43
nforum_votes                    6
nforum_endorsed                33
nforum_threads                 66
nforum_comments                 5
nforum_pinned                   3
roles                         107
nprogcheck                    578
nproblem_check                213
nforum_events                   5
mode                            6
is_active                   22286
cert_created_date           21901
cert_modified_date            608
cert_status                   607

In [338]:
stats_NaN(df_clean)

NaN Ratio
course_id                       0
user_id                         0
registered                      0
viewed                0.000270603
explored              0.000498877
certified             0.000757386
completed              0.00102497
ip                     0.00126231
cc_by_ip               0.00155408
countryLabel           0.00181712
continent              0.00207563
city                   0.00235379
region                 0.00264405
subdivision            0.00291767
postalCode             0.00316258
un_major_region        0.00343167
un_economic_group      0.00372192
un_developing_nation    0.0039759
un_special_region      0.00421929
latitude               0.00446117
longitude              0.00468339
LoE                      0.004954
YoB                    0.00520948
gender                 0.00549067
grade                  0.00576883
passing_grade          0.00600617
start_time             0.00624805
first_event            0.00652017
last_event              0.0067424
nevents                0.00698427
ndays_act              0.00722464
nplay_video            0.00749978
nchapters               0.0077462
nforum_posts           0.00801226
nforum_votes           0.00826473
nforum_endorsed        0.00856707
nforum_threads         0.00882407
nforum_comments        0.00909619
nforum_pinned          0.00936528
roles                  0.00962076
nprogcheck             0.00989892
nproblem_check         0.00989892
nforum_events          0.00989892
mode                   0.00989892
is_active              0.00989892
cert_created_date      0.00989892
cert_modified_date     0.00989892
cert_status            0.00989892

In [28]:
display_df(df_clean,200)

course_id  user_id registered viewed explored  \
1       HarvardX/PH525.1x/1T2018     7940      False           False   
2       HarvardX/PH525.1x/1T2018     7940      False           False   
3       HarvardX/PH525.1x/1T2018     7940      False           False   
4       HarvardX/PH525.1x/1T2018     7940      False           False   
5       HarvardX/PH525.1x/1T2018     7940      False           False   
6       HarvardX/PH525.1x/1T2018     7940      False           False   
7       HarvardX/PH525.1x/1T2018     7940      False           False   
8       HarvardX/PH525.1x/1T2018     7940      False           False   
9       HarvardX/PH525.1x/1T2018     7940      False           False   
10      HarvardX/PH525.1x/1T2018     7940      False           False   
11      HarvardX/PH525.1x/1T2018     7940      False           False   
12      HarvardX/PH525.1x/1T2018     7940      False           False   
13      HarvardX/PH525.1x/1T2018     7940      False           False   
14      HarvardX/PH525.1x/1T2018     7940      False           False   
15      HarvardX/PH525.1x/1T2018     7940      False           False   
16      HarvardX/PH525.1x/1T2018    21193       True   True    False   
17      HarvardX/PH525.1x/1T2018    21193       True   True    False   
18      HarvardX/PH525.1x/1T2018    21193       True   True    False   
19      HarvardX/PH525.1x/1T2018    21193       True   True    False   
20      HarvardX/PH525.1x/1T2018    21193       True   True    False   
21      HarvardX/PH525.1x/1T2018    21193       True   True    False   
22      HarvardX/PH525.1x/1T2018    21193       True   True    False   
23      HarvardX/PH525.1x/1T2018    21193       True   True    False   
24      HarvardX/PH525.1x/1T2018    21193       True   True    False   
25      HarvardX/PH525.1x/1T2018    21193       True   True    False   
26      HarvardX/PH525.1x/1T2018    21193       True   True    False   
27      HarvardX/PH525.1x/1T2018    21193       True   True    False   
28      HarvardX/PH525.1x/1T2018    21193       True   True    False   
29      HarvardX/PH525.1x/1T2018    21193       True   True    False   
30      HarvardX/PH525.1x/1T2018    21193       True   True    False   
31      HarvardX/PH525.1x/1T2018    21193       True   True    False   
32      HarvardX/PH525.1x/1T2018    21193       True   True    False   
33      HarvardX/PH525.1x/1T2018    21193       True   True    False   
34      HarvardX/PH525.1x/1T2018    21193       True   True    False   
35      HarvardX/PH525.1x/1T2018    21193       True   True    False   
36      HarvardX/PH525.1x/1T2018    21193       True   True    False   
37      HarvardX/PH525.1x/1T2018    21193       True   True    False   
38      HarvardX/PH525.1x/1T2018    27938       True   True    False   
39      HarvardX/PH525.1x/1T2018    27938       True   True    False   
40      HarvardX/PH525.1x/1T2018    27938       True   True    False   
41      HarvardX/PH525.1x/1T2018    27938       True   True    False   
42      HarvardX/PH525.1x/1T2018    27938       True   True    False   
43      HarvardX/PH525.1x/1T2018    27938       True   True    False   
44      HarvardX/PH525.1x/1T2018    27938       True   True    False   
45      HarvardX/PH525.1x/1T2018    27938       True   True    False   
46      HarvardX/PH525.1x/1T2018    27938       True   True    False   
47      HarvardX/PH525.1x/1T2018    27938       True   True    False   
48      HarvardX/PH525.1x/1T2018    28454       True  False    False   
49      HarvardX/PH525.1x/1T2018    28454       True  False    False   
50      HarvardX/PH525.1x/1T2018    28454       True  False    False   
51      HarvardX/PH525.1x/1T2018    28454       True  False    False   
52      HarvardX/PH525.1x/1T2018    28454       True  False    False   
53      HarvardX/PH525.1x/1T2018    28454       True  False    False   
54      HarvardX/PH525.1x/1T2018    28454       True  False    False   
55      HarvardX/PH525.1x/1T2018    28454       True  False    

The first perceivable error in the dataset occurs under the completed field. It often has the IP in it. Therefore, one of the early fields is the source of the problem. Perhaps one field is missing. However we are finding it difficult to determine which one. Sometimes, `viewed` is empty, sometimes `explored` is. Often we see `registered` equal to False, which also makes no sense as in order to be a row in the dataset that user needs to be registered for the course. Initially we are thinking that whenever `registered` is False, that it should really be True and that is the source of the missing field and therefore the cause of the shift. However, we cant be certain. Sometimes `registered` is indeed True. `completed` is an important field for researchers of the edX dataset, so we most definitely must repair this column, and we can likely assume that one of the previous columns, probably `certified` contains the values for that column. But which column is missing? The four fields previous `completed` are all booleans, so it is difficult to identify where the problem originates. We know for sure that if the `ip` is in the completed column we can shift the datapoints starting in `completed` over to the right one. We have decided that the source of the shift is the `registered` column. It is the least important for researchers as it is redundant and should always be True. Therefore we start the shift over from there and fill the value with True.

We target the `completed` column for values that look like IP addresses. In this case, we are looking for values that would not be the acceptable values in the `completed` column. The `completed` column should have either True, False, or perhaps None or empty string - anything that is not these is likely an IP address. **We discovered later that the empty string is a source of a shift and remedied that.** We then shift everything beginning at `registered` over as that is a reasonable assumption as to what the missing column is and we will then fill that column with Trues as everyone in the database is registered for the class. We do recognize that any of the other boolean fields could have been the source of the shift like `viewed`, `explored`, `certified`, or even `completed` (though we think this is less likely given these fields seem more important for analysis). A cursory glance at the data shows that `explored` is often empty - could this be the source of the problem? We think not, as we remember many NaN values in the `explored` column in the previous pset.

In [454]:
#gets rows that look like the IP or the empty string is in the completed field 
#and shifts them
df_shift = df_clean[~df_clean['completed'].isin(['True','False', None])]\
[df_clean.columns[2:]].shift(1, axis=1)
df_shift.registered = 'True'

In [455]:
df_clean.loc[~df_clean['completed'].isin(['True','False', None]),\
            df_clean.columns[2:]] = df_shift

In [456]:
df_clean[df_clean['user_id']=='2014-09-15 17:57:23']

course_id              user_id  registered viewed explored certified  \
652240   6:36:03  2014-09-15 17:57:23  notpassing            None      None   

       completed    ip cc_by_ip countryLabel     ...     nforum_pinned roles  \
652240      None  None     None         None     ...              None  None   

       nprogcheck nproblem_check nforum_events  mode is_active  \
652240       None           None          None  None      None   

       cert_created_date cert_modified_date cert_status  
652240              None               None        None  

[1 rows x 48 columns]

This row is identified in our `fix_types` function as a shift, but it is complete garbage, so we will just drop it.

In [457]:
df_clean = df_clean.drop(652240)

We have successfully fixed the shifting in the dataset! Our function below proves it as it does not identify any shifts.

### Fixing Data Types and Handling Missingness

Now that we have fixed the shifting, it will be more efficient to fix the datatypes. They are incorrect in multiple ways. For example, things that say False are not really the boolean value for false, but rather the string. In fact everything is a string. We will correct this in a systematic way. We will try to best follow the guidelines outlined here: https://elitedatascience.com/data-cleaning. We will have 2 functions. One for handling the fields that are supposed to be strings and the other for the other types. `fix_types_strings` will be like pandas replace function and deal with finding the empty strings. The other function `fix_types` will take a dictionary of column names to datatypes and run through the columns correcting the datatypes. Columns not specified in the dictionary will be left as strings.

The dictionary for chaning the non-string types is below. Most of these type conversions are self explanatory. But some of these choices in type conversions and in conversions I dont make require explanation. I am leaving `postalCode` as a string: this is because not all postal codes are numbers, some contain letters like in Great Britain. For `latitude` and `longitude` a float is enough space to hold their values, a double is not necessary. Many of the fields require datetime format which will be handled in the `fix_types` function. Lastly, `is_active` is described as an integer 1 or 0 in the documentation (and not a boolean) so we will treat it as such. The dictionary is equipped with the default value in the case that a datapoint is missing. Our `fix_types` function will utilize this.

In [458]:
missing_int = 0
missing_float = 0
missing_bool = False
missing_datetime = datetime.min
type_dict = {
    'user_id' : (int, missing_int),
    'registered' : (bool, missing_bool),
    'viewed' : (bool, missing_bool),
    'explored' : (bool, missing_bool),
    'certified' : (bool, missing_bool),
    'completed' : (bool, missing_bool),
    'latitude' : (float, missing_float),
    'longitude' : (float, missing_float), 
    'YoB' : (int, missing_int),
    'grade' : (float, missing_float),
    'passing_grade' : (float, missing_float),
    'start_time' : ('datetime', missing_datetime),
    'first_event' : ('datetime', missing_datetime),
    'last_event' : ('datetime', missing_datetime),
    'nevents' : (int, missing_int),
    'ndays_act' : (int, missing_int),
    'nplay_video' : (int, missing_int),
    'nchapters' : (int, missing_int),
    'nforum_posts' : (int, missing_int),
    'nforum_votes' : (int, missing_int),
    'nforum_endorsed' : (int, missing_int),
    'nforum_threads' : (int, missing_int),
    'nforum_comments' : (int, missing_int),
    'nforum_pinned' : (int, missing_int),
    'nprogcheck' : (int, missing_int),
    'nproblem_check' : (int, missing_int),
    'nforum_events' : (int, missing_int),
    'is_active' : (int, missing_int),
    'cert_created_date' : ('datetime', missing_datetime),
    'cert_modified_date' : ('datetime', missing_datetime)
}

The `fix_types` function will be equipped to deal with missingness in the columns provided to the function. Following the guidelines from above, we will "flag and fill", meaning we will put in a default value like 0 and add a column indicating the a value in that field was missing. We will deal with the missingness in the string values after.

In [459]:
dt_format = "%Y-%m-%d %H:%M:%S"
dt_format_catch = "%Y-%m-%d %H:%M:%S.%f"

In [460]:
"""
Goes through columns specified in the dictionary (the keys) and casts them as the 
appropriate type (the value). The dict will only contain fields that need to be 
changed from string, as everything else should be left as a string. For fields of 
datetime format, the input variables dt_format is the format provided to strptime,
dt_format_catch is the alternative datetime format in the case dt_format caused an 
error. This function 'flags and fills', meaning it fills missing values with the 
default specified in the dictionary and adds columns to the dataframe to indicate
the missingness of fields.
"""
def fix_types(df, type_dict, dt_format, dt_format_catch):
    #the function passed to the apply on the column
    def fix_type(s, type_value, missing_value, dt_format, dt_format_catch):
        if pd.isnull(s) or s=='' or s==' ' or s=='none':
            return missing_value
        if type_value != 'datetime':
            try:
                return type_value(s)
            except ValueError as v: #handles junk datapoints
                #helps identify other shifts
                print('potential shift: '+s) #if correct, should never print
                return missing_value
        elif type_value == 'datetime':
            try:
                return datetime.strptime(s, dt_format)
            except ValueError as v:
                return datetime.strptime(s, dt_format_catch)
    for key, (type_value, missing_value) in type_dict.items():
        print('Cleaning ' + str(key) + '...')
        df[key] = df[key].apply(lambda s : fix_type(s, type_value, missing_value,
                                                   dt_format, dt_format_catch))
        #check if contains missing_value and add column
        if missing_value in df[key].values:
            df['missing_'+str(key)] = df[key]==missing_value
    print('Finished!')
    return df

We have written the function to help us find more shifts. For example, we have just identified that the empty string in the `completed` field is the source of a shift. We will go back and fix that now.

In [461]:
df_clean = fix_types(df_clean, type_dict, dt_format, dt_format_catch)

Cleaning user_id...
Cleaning registered...
Cleaning viewed...
Cleaning explored...
Cleaning certified...
Cleaning completed...
Cleaning latitude...
Cleaning longitude...
Cleaning YoB...
Cleaning grade...
Cleaning passing_grade...
Cleaning start_time...
Cleaning first_event...
Cleaning last_event...
Cleaning nevents...
Cleaning ndays_act...
Cleaning nplay_video...
Cleaning nchapters...
Cleaning nforum_posts...
Cleaning nforum_votes...
Cleaning nforum_endorsed...
Cleaning nforum_threads...
Cleaning nforum_comments...
Cleaning nforum_pinned...
Cleaning nprogcheck...
Cleaning nproblem_check...
Cleaning nforum_events...
Cleaning is_active...
Cleaning cert_created_date...
Cleaning cert_modified_date...
Finished!


Now that we have dealt with missingness in the non-string types, we will handle the string types. The guidelines recommend filling in missing categoricals (in our case all of the strings) with "missing". So we shall do that.

In [469]:
df_clean = df_clean.fillna('missing')

In [470]:
#missing_list=['',' ']
#df_clean = df_clean.replace(missing_list, 'missing')

Pandas replace is incredibly slow. So we will write are own function.

In [471]:
str_cols = list(set(df_clean.columns) - set(type_dict.keys()))
str_cols = [s for s in str_cols if not s.startswith('missing')]

In [472]:
"""
Fills in missing values identified by the missing_list with the proper missing_value.
Pandas fillna handles the other missing value identifers like NaN and None.
"""
def fix_types_strings(df, cols, missing_list=['',' '], missing_value='missing'):
    for col in cols:
        df[col] = df[col].map(lambda s: missing_value if s in missing_list else s)
    return df

In [473]:
df_clean = fix_types_strings(df_clean, str_cols)

We have cleaned and dealt with missingness in the fields that are numerical boolean, and datetime, and properly flagged them. Now we are finished dealing with missingness and fixing the datatypes.

### Getting Rid of Duplicates

### Testing

In [96]:
def is_ip(s):
    return (len(s.split('.'))==4)

In [ ]:
def check_types(df):
    

## Dealing with Corruption (Part 2)

In [168]:
#garbage, will have to deal with
display_df(df_clean[df_clean['completed'].isnull()])

course_id  user_id registered viewed explored  \
758     HarvardX/PH525.1x/1T2018   529339       True   True     None   
1035    HarvardX/PH525.1x/1T2018   866059      False            None   
3495    HarvardX/PH525.1x/1T2018  3524821       True   None     None   
3828    HarvardX/PH525.1x/1T2018  4030650       True   True     None   
4925    HarvardX/PH525.1x/1T2018  5498643       True   True    False   
...                          ...      ...        ...    ...      ...   
653964    HarvardX/CS50x/2014_T1  1842651       True   True    False   
656251    HarvardX/CS50x/2014_T1  1876271       True  False            
657242    HarvardX/CS50x/2014_T1  1893442       True   True    False   
658870    HarvardX/CS50x/2014_T1  1916566       True  False            
659931    HarvardX/CS50x/2014_T1  1923170       True  False            

       certified completed    ip cc_by_ip countryLabel continent  city region  \
758         None      None  None     None         None      None  None   None   
1035        None      None  None     None         None      None  None   None   
3495        None      None  None     None         None      None  None   None   
3828        None      None  None     None         None      None  None   None   
4925       False      None  None     None         None      None  None   None   
...          ...       ...   ...      ...          ...       ...   ...    ...   
653964      None      None  None     None         None      None  None   None   
656251      None      None  None     None         None      None  None   None   
657242     False      None  None     None         None      None  None   None   
658870     False      None  None     None         None      None  None   None   
659931     False      None  None     None         None      None  None   None   

       subdivision postalCode un_major_region un_economic_group  \
758           None       None            None              None   
1035          None       None            None              None   
3495          None       None            None              None   
3828          None       None            None              None   
4925          None       None            None              None   
...            ...        ...             ...               ...   
653964        None       None            None              None   
656251        None       None            None              None   
657242        None       None            None              None   
658870        None       None            None              None   
659931        None       None            None              None   

       un_developing_nation un_special_region latitude longitude   LoE   YoB  \
758                    None              None     None      None  None  None   
1035                   None              None     None      None  None  None   
3495                   None              None     None      None  None  None   
3828                   None              None     None      None  None  None   
4925                   None              None     None      None  None  None   
...                     ...               ...      ...       ...   ...   ...   
653964                 None              None     None      None  None  None   
656251                 None              None     None      None  None  None   
657242                 None              None     None      None  None  None   
658870                 None              None     None      None  None  None   
659931                 None              None     None      None  None  None   

       gender grade passing_grade start_time first_event last_event nevents  \
758      None  None          None       None        None       None    None   
1035     None  None          None       None        None       None    None   
3495     None  None          None       None        None       None    None   
3828     None  None          None       None        None       None    None   
4925     None  None     

A quick find shows that certain students with failing grades are marked as `completed`. This could a sign of corruption. Or perhaps something else that we do not understand. We are considering adding a field `corrupt` that would mark the corrupt lines and marking rows like this as such.

In [380]:
display_df(df_clean[(df_clean['gender'] < df_clean['grade']) & (df_clean['certified']=="True")],200)

TypeError: '<' not supported between instances of 'str' and 'float'